libs/utils/src/http/endpoint.rs

```diff
@@ -328,6 +336,122 @@ pub async fn prometheus_metrics_handler(_req: Request<Body>) -> Result<Response<
+        return Err(ApiError::InternalServerError(anyhow!(
+            "heap profiling not enabled"
+        )));
+    }
+
+    // Take and return the profile.
+    match format {
+        Format::Jemalloc => {
+            // NB: file is an open handle to a tempfile that's already deleted.
+            let file = tokio::task::spawn_blocking(move || prof_ctl.dump())
+                .await
+                .map_err(|join_err| ApiError::InternalServerError(join_err.into()))?
+                .map_err(ApiError::InternalServerError)?;
+            let stream = ReaderStream::new(tokio::fs::File::from_std(file));
+            Response::builder()
+                .status(200)
+                .header(CONTENT_TYPE, "application/octet-stream")
+                .header(CONTENT_DISPOSITION, "attachment; filename=\"heap.dump\"")
+                .body(Body::wrap_stream(stream))
+                .map_err(|err| ApiError::InternalServerError(err.into()))
+        }
+
+        Format::Pprof => {
+            let data = tokio::task::spawn_blocking(move || prof_ctl.dump_pprof())
+                .await
+                .map_err(|join_err| ApiError::InternalServerError(join_err.into()))?
+                .map_err(ApiError::InternalServerError)?;
+            Response::builder()
+                .status(200)
+                .header(CONTENT_TYPE, "application/octet-stream")
+                .header(CONTENT_DISPOSITION, "attachment; filename=\"heap.pb\"")
+                .body(Body::from(data))
+                .map_err(|err| ApiError::InternalServerError(err.into()))
+        }
+
+        Format::Svg => {
+            let tempfile = tokio::task::spawn_blocking(move || {
+                let mut file = prof_ctl.dump()?;
+                // dump() returns a file that's already been deleted from the file system, but with
+                // an open file handle to read from. jeprof_svg() needs a path to a file in the file
+                // system. We therefore copy the file to another temp file we can get the path to.
+                let mut tempfile = NamedUtf8TempFile::new()?;
+                std::io::copy(&mut file, &mut tempfile)?;
+                Ok(tempfile)
+            })
+            .await
+            .map_err(|join_err| ApiError::InternalServerError(join_err.into()))?
+            .map_err(ApiError::InternalServerError)?;
+            let svg = jeprof_svg(tempfile.path())
+                .await
+                .map_err(ApiError::InternalServerError)?;
+            Response::builder()
+                .status(200)
+                .header(CONTENT_TYPE, "image/svg+xml")
+                .body(Body::from(svg))
+                .map_err(|err| ApiError::InternalServerError(err.into()))
+        }
+    }
+}
+
+/// Generates an SVG for the given jemalloc profile file. Requires Perl installed.
+///
+/// TODO: jeprof is really slow (~20 seconds) because of addr2line.
+pub async fn jeprof_svg(path: &Utf8Path) -> anyhow::Result<Vec<u8>> {
```

Comment:
Do we really want this as a service route? This should be an offline task.

Reply:
FWIW, this is lifted from tikv: https://github.com/tikv/tikv/blob/b6252e9b66657f6767301ca183dbf7af9265a849/src/server/status_server/mod.rs#L153

It's a pretty common practice.

Reply:
Wdym? Go's pprof web server can take CPU/heap profiles and display them as interactive flamegraphs, charts, text etc in a browser. This is all done on the server.

The same service can also be run locally via go tool pprof, but it's available through the pprof HTTP service as well.

Reply:
Sorry, I don't see where this happens in Go's stdlib. Anyway, if you feel this is useful to try out and worth the risk of running the jeprof script, then go ahead. Ideally, though, Grafana should provide us with useful graphs. The service port is not easily accessible anyway.

Reply:
> Sorry, I don't see where this happens in Go's stdlib. 

You're right, my mistake -- the systems I've used explicitly wired up the pprof web UI.

> if you feel this is useful to try out and worth the risk of running the jeprof script, then go ahead. Ideally, though, Grafana should provide us with useful graphs.

I'd prefer to get readable profile output in a single step. This is faster during incidents, and also makes them accessible to non-developers (e.g. SREs). But as you say, Grafana would presumably also provide this, at least for production and staging clusters.

For CPU profiles in https://github.com/neondatabase/neon/pull/9764, `pprof-rs` can natively emit flamegraph SVGs, so I don't see a compelling reason not to support it. But since heap profiles appear to be a lot more hassle it might not be worth it.

In any case, I think we'll have to do server-side profile symbolization in order to use these with Grafana. Let's solve that first, and then we can see what the marginal cost of other output formats would be. For now, let's remove jeprof here to unblock this, and I'll revisit this when I have time.